File: dsa/concepts/16_design/python/l981_time_based_key_value_store.py
# Problem: https://leetcode.com/problems/time-based-key-value-store/
# Idea:
# Store for each key a list of (value, timestamp) pairs in increasing timestamp order.
# For get(), use Binary Search to find the value with the largest timestamp <= given timestamp.
#
# set  -> O(1)
# get  -> O(log n)
# Space -> O(n)

class TimeMap:

    def __init__(self):
        # key -> list of (value, timestamp)
        self.map = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.map:
            self.map[key] = []
        self.map[key].append((value, timestamp))

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.map:
            return ""

        values = self.map[key]
        low, high = 0, len(values) - 1
        res = ""

        # Binary search for closest time <= timestamp
        while low <= high:
            mid = (low + high) // 2
            val, time = values[mid]

            if time <= timestamp:
                res = val
                low = mid + 1
            else:
                high = mid - 1

        return res


# ---------------- Test (same as Java main) ----------------

obj = TimeMap()

obj.set("foo", "bar", 1)
print(obj.get("foo", 1))   # bar
print(obj.get("foo", 3))   # bar
obj.set("foo", "bar2", 4)
print(obj.get("foo", 4))   # bar2
print(obj.get("foo", 5))   # bar2

File: dsa/topics/16_design/java/L981_TimeBasedKeyValueStore.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class L981_TimeBasedKeyValueStore {

	public static void main(String[] args) {
		TimeMap obj = new TimeMap();

		obj.set("foo","bar",1);
		System.out.println(obj.get("foo",1));
		System.out.println(obj.get("foo",3));
		obj.set("foo","bar2",4);
		System.out.println(obj.get("foo",4));
		System.out.println(obj.get("foo",5));
	}
}
// Problem: https://leetcode.com/problems/time-based-key-value-store/
// Idea: Use BS to find closest time to target time
// Time complexity: O(logn)
// Follow ups:
//  i) how to delete a key
//  ii) how to implement a multi thread version
class TimeMap{
	HashMap<String, ArrayList<Pair>> map;   // [key = (value, timestamp)]
    public TimeMap() {
        map = new HashMap<>();
    }

    public void set(String key, String value, int timestamp) {
        map.putIfAbsent(key, new ArrayList<>());
        map.get(key).add(new Pair(value, timestamp));
    }

    public String get(String key, int timestamp) {
        List<Pair> list = map.getOrDefault(key, new ArrayList<>());
        int low=0, high=list.size()-1;
        String res = "";
        while(low <= high){
            int mid = (low+high)/2;
            Pair pair = list.get(mid);
            if(pair.time <= timestamp){
                res = pair.val;
                low = mid+1;
            }else{
                high = mid-1;
            }
        }

        return res;
    }
}

class Pair{
    String val;
    int time;

    Pair(String val, int time){
        this.val = val;
        this.time = time;
    }
}
